### PR TITLE
raftstore-v2: flush memtable before proposing split

### LINKED ISSUE #14447: [Dynamic Regions] Split can significantly impacts write QPS
## Bug Report
When there're region split, the write qps is impacted a lot.  This is because the split involves memtable flush in raftstore threads which could block raftstore messages from being processed

### What version of TiKV are you using?

nightly

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Close #14447

What's Changed:

```commit-message
flush memtable before proposing split
```
Split calls `checkpoint` API of RocksDB which can easily cost more than 500ms.
As the main expensive operation of the `checkpoint `is the memtable flush, this PR schedule memtable flush before proposing the split.
So, when split is to be proposed, schedule a memtable flush task to the relevant worker, and at the same time, send raft messages with extra msg to followers to let them flush their memtables.
After the flush task finishes, the split can be proposed, and this time, `checkpoint` will be much more lightweight.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
Please add a release note.

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

If you don't think this PR needs a release note then fill it with None.
If this PR will be picked to release branch, then a release note is probably required.
```

### CODE REVIEW

File: tests/integrations/raftstore/test_split_region.rs [hunk 1]
```diff
@@ -145,7 +145,15 @@ fn test_server_split_region_twice() {
     rx1.recv_timeout(Duration::from_secs(5)).unwrap();
 }
 
-fn test_auto_split_region<T: Simulator>(cluster: &mut Cluster<T>) {
+#[test_case(test_raftstore::new_node_cluster)]
+#[test_case(test_raftstore::new_server_cluster)]
+#[test_case(test_raftstore::new_incompatible_node_cluster)]
```
Comment (Connor1996) [inline tests/integrations/raftstore/test_split_region.rs:150]: Could you depict the difference among node_cluster, server_cluster and incompatible_node_cluster in comment of somewhere

Comment (SpadeA-Tang) [reply tests/integrations/raftstore/test_split_region.rs:150]: The modification here is just reuse v1's relevant tests for v2.

Comment (SpadeA-Tang) [reply tests/integrations/raftstore/test_split_region.rs:150]: Fine, I will add them.

File: components/raftstore-v2/src/worker/tablet_flush.rs [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    }
+
+    fn flush_tablet(
+        &mut self,
+        region_id: u64,
+        req: Option<RaftCmdRequest>,
+        is_leader: bool,
+        ch: Option<CmdResChannel>,
+    ) {
+        if let Some(mut cache) = self.tablet_registry.get(region_id) {
```
Comment (bufferflies) [inline components/raftstore-v2/src/worker/tablet_flush.rs:61]: maybe you can decrease the `if else`

File: components/raftstore-v2/src/worker/tablet_flush.rs [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        is_leader: bool,
+        ch: Option<CmdResChannel>,
+    },
+}
+
+impl Display for Task {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Task::TabletFlush { region_id, .. } => {
+                write!(f, "Flush tablet before split for region {}", region_id)
```
Comment (bufferflies) [inline components/raftstore-v2/src/worker/tablet_flush.rs:29]: only for split ?

File: components/raftstore-v2/src/worker/tablet_flush.rs [hunk 4]
```diff
@@ -0,0 +1,111 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
```
Comment (bufferflies) [inline components/raftstore-v2/src/worker/tablet_flush.rs:1]:
```suggestion
// Copyright 2023 TiKV Project Authors. Licensed under Apache-2.0.
```

File: components/raftstore-v2/src/worker/tablet_flush.rs (hunk 3 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/worker/tablet_flush.rs:29]: Now, it is.

File: components/raftstore-v2/src/worker/tablet_flush.rs (hunk 4 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/worker/tablet_flush.rs:1]: Good catch.

**Review by Connor1996 — COMMENTED** [review]
`After the flush task finishes, the split can be proposed, and this time, checkpoint will be much more lightweight.`

What if there are some writes before the split is proposed and after the flush is triggered? Then the split still needs to trigger flush.

**Comment by SpadeA-Tang** [comment]
> What if there are some

Yes, but it is much more lightweight. At majority cases, it can reduce flush time from serveral handred ms to dozens of ms.

File: components/raftstore-v2/src/worker/tablet_flush.rs (hunk 2 above)
Comment (tonyxuqqi) [reply components/raftstore-v2/src/worker/tablet_flush.rs:61]: +1.  Too deep if block.

**Review by tonyxuqqi — APPROVED** [review]
A minor code style issue. Other LGTM

File: components/raftstore-v2/src/operation/command/admin/compact_log.rs [hunk 5]
```diff
@@ -113,7 +113,7 @@ impl<'a, EK: KvEngine, ER: RaftEngine, T: Transport> PeerFsmDelegate<'a, EK, ER,
 
 impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
     // Mirrors v1::on_raft_gc_log_tick.
-    fn maybe_propose_compact_log<T>(
+    fn maybe_propose_compact_log<T: Transport>(
```
Comment (bufferflies) [inline components/raftstore-v2/src/operation/command/admin/compact_log.rs:116]: why change it?

File: components/raftstore-v2/src/operation/command/admin/mod.rs [hunk 6]
```diff
@@ -118,7 +126,81 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
...
+                            ) {
+                                error!(
+                                    self.logger,
+                                    "Fail to schedule flush task";
+                                    "err" => ?e,
+                                )
+                            }
+
+                            let peers = self.region().get_peers().to_vec();
+                            for p in peers {
```
Comment (bufferflies) [inline components/raftstore-v2/src/operation/command/admin/mod.rs:175]: Does it needs to send it for all peer includes witness or tiflash?

File: components/raftstore-v2/src/worker/tablet_flush.rs [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+            .tablet_registry
+            .get(region_id)
+            .map(|mut cache| cache.latest().cloned()) else {return};
+        let now = Instant::now();
+        tablet.flush_cfs(DATA_CFS, true).unwrap();
+        let elapsed = now.saturating_elapsed();
+        info!(
+            self.logger,
+            "Flush memtable time consumes";
+            "region_id" =>  region_id,
```
Comment (bufferflies) [inline components/raftstore-v2/src/worker/tablet_flush.rs:71]:
```suggestion
            "region_id" => region_id,
```

File: components/raftstore-v2/src/operation/command/admin/compact_log.rs (hunk 5 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/command/admin/compact_log.rs:116]: it calls on_admin_command, which needs to sends msgs now.

File: components/raftstore-v2/src/operation/command/admin/mod.rs (hunk 6 above)
Comment (SpadeA-Tang) [reply components/raftstore-v2/src/operation/command/admin/mod.rs:176]: Good catch. I think filtering witness and non-voter is enough.

File: components/raftstore-v2/src/batch/store.rs [hunk 8]
```diff
@@ -497,19 +498,24 @@ struct Workers<EK: KvEngine, ER: RaftEngine> {
...
     purge: Option<Worker>,
+    tablet_flush: Worker,
 
     // Following is not maintained by raftstore itself.
     background: Worker,
 }
 
 impl<EK: KvEngine, ER: RaftEngine> Workers<EK, ER> {
     fn new(background: Worker, pd: LazyWorker<pd::Task>, purge: Option<Worker>) -> Self {
+        let tablet_flush = WorkerBuilder::new("tablet_flush-worker")
```
Comment (Connor1996) [inline components/raftstore-v2/src/batch/store.rs:509]: mixed use of `_` and `-`

**Review by Connor1996 — APPROVED** [review]
rest LGTM

**Review by bufferflies — APPROVED** [review]

**Comment by BusyJay** [comment]
Sending the flush command around is not accurate and leaks extraction of memtable. When FreezeAndClone is implemented, the protocol becomes obsolete. Perhaps you can offload the checkpoint operation to other threads instead, so apply thread will not be blocked. That is leader continues to trigger flush when proposing split to avoid waiting too long in apply thread, every peer offloads checkpoint/split to other threads to avoid blocking apply thread.

File: components/raftstore-v2/src/operation/command/admin/mod.rs [hunk 9]
```diff
@@ -122,7 +131,84 @@ impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
...
+                                if p == *self.peer()
+                                    || p.get_role() != PeerRole::Voter
```
Comment (zhangjinpeng1987) [inline components/raftstore-v2/src/operation/command/admin/mod.rs:182]: should be == voter?

File: components/raftstore-v2/src/worker/tablet_flush.rs [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+            "region_id" => region_id,
+            "duration" => ?elapsed,
+            "is_leader" => is_leader,
+        );
+
+        if !is_leader {
+            return;
+        }
+
+        let mut req = req.unwrap();
```
Comment (zhangjinpeng1987) [inline components/raftstore-v2/src/worker/tablet_flush.rs:81]: Need to check if the original request is a split request.

Comment (SpadeA-Tang) [reply components/raftstore-v2/src/worker/tablet_flush.rs:81]: Good suggestion.

**Pull request merged**